feat: add `ndarray/base/binary-reduce-strided1d-dispatch`

gururaj1512 opened the pull request:
Resolves None

## Description

> What is the purpose of this pull request?

This pull request:

-   feat: add `ndarray/base/binary-reduce-strided1d-dispatch`

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves None

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

gururaj1512 committed:
feat: add `ndarray/base/binary-reduce-strided1d-dispatch`

gururaj1512 committed:
fix: minor bug

gururaj1512 committed:
fix: correct comments in examples and README

kgryte commented:
/stdlib merge

stdlib-bot committed:
Merge remote-tracking branch 'upstream/develop' into ndarray-base-binary-reduce-strided1d-dispatch

kgryte commented:
This PR is dependent on https://github.com/stdlib-js/stdlib/pull/7904.

kgryte commented:
/stdlib merge

stdlib-bot committed:
Merge remote-tracking branch 'upstream/develop' into ndarray-base-binary-reduce-strided1d-dispatch

kgryte commented:
@gururaj1512 Now that #7904 is merged, is there anything else you want to do to this PR before review?

Note that I changed some of the logic in #7904. In particular, for each input ndarray, to determine the casting data type, one would do

```
var idt1 = getDType( x1 );
var idt2 = getDType( x2 );

var dt1 = binaryCastingDataType( idt1, idt2, odt, policy );
var dt2 = binaryCastingDataType( idt2, idt1, odt, policy );
```

Here, the first argument corresponds to the array dtype that you want to figure out the casting data type and the second argument corresponds to the "additional"/other input ndarray data type.

kgryte commented:
The lint failure seems to be due to a recently added lint rule which is buggy.

cc @Planeshifter

kgryte commented:
/stdlib merge

stdlib-bot committed:
Merge remote-tracking branch 'upstream/develop' into ndarray-base-binary-reduce-strided1d-dispatch

kgryte commented:
As the REPL text lint failure is expected, I'll go ahead and merge.

A few comments @headlessNode:

1. Policies have been removed. They were not being used, so it is pointless to have a `policies` argument.
2. You still needed to support separate input and output dtype arguments. You seemed to fundamentally misunderstand how `idtypes` was intended to work. In particular, that array is intended to place dtype constraints on any, including additional, input ndarray arguments. That is a separate concern from the dtype constraints placed on the input/output ndarray which is the target of the nullary strided function.

gururaj1512 committed:
fix: cast to proper dtype

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

gururaj1512 commented:
@kgryte, I have refactored the code as per changed #7904.
Here, inputs cast individually for each ndarray instead of to common type.
Ready for review

kgryte committed:
chore: clean-up

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: passed
  - task: lint_repl_help
    status: skipped
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

kgryte reviewed (approved):
LGTM after clean-up.

kgryte commented:
As the CI lint failure is expected, I'll go ahead and merge.
